### PR TITLE
Autograph decorated functions with tf.map_fn

### PR DESCRIPTION
**Pull request by bhack**
This is a test to verify the failure of the case described in https://github.com/tensorflow/tensorflow/issues/26091#issuecomment-628304623

/cc @dthkao

### CODE REVIEW

File: tensorflow/python/autograph/impl/api_test.py
Comment (mdanatg) [inline tensorflow/python/autograph/impl/api_test.py:295]: This statement has no effect - perhaps you meant it to be a decorator?
At any rate, converted_call is not intended to be a decorator, but you can still use it as x = api.converted_call(test_call_partial, ...)

Comment (bhack) [reply tensorflow/python/autograph/impl/api_test.py]: Yes sorry was a wrong cut and paste from another experiment. It was `api.convert(recursive=DEFAULT_RECURSIVE)`

Comment (mdanatg) [reply tensorflow/python/autograph/impl/api_test.py]: Got it, noprob. I think we still need the `@` to make it a decorator: `@api.convert(...)`?

Comment (bhack) [reply tensorflow/python/autograph/impl/api_test.py]: Yes i fixed the typo

**Review by mdanatg — DISMISSED** [review]

**Comment by bhack** [comment]
Mhh.. why test Is passing?

**Comment by mdanatg** [comment]
I think the issue was fixed a while ago, but the issue ID was never tagged by the fix. I definitely expect the test to pass. Have you seen recent cases when it doesn't work?

**Comment by bhack** [comment]
We had problems yesterday with 2.2.0 see https://github.com/tensorflow/addons/pull/1830#issuecomment-628283237 and the [colab](https://github.com/tensorflow/addons/pull/1830#issuecomment-628264997) in that ticket.

**Comment by bhack** [comment]
And also with `tf-nightly`

**Comment by mdanatg** [comment]
I see. Do you have a small reproducer in a colab, using @tf.function as decorator? Something simple like this test would do. It might be caused by keyword arguments, if I read through the comments correctly.

**Comment by bhack** [comment]
```
def test_function(x):
    cond = tf.constant(-1)

    if cond == 0:
        result = x
    else:
        result = x

    return result

@tf.function
def partial_call(x):
    fn = partial(test_function)
    tf.map_fn(fn, x)
```

```
X = tf.random.uniform(shape=(10, 224, 224, 3))
partial_call(X)
```

**Comment by mdanatg** [comment]
I don't think that's caused by `partial`. The problem is that `map_fn` doesn't call autograph (see below). That's definitely a bug - feel free to file an issue, and if you're interested to help fixing it, I can give a few pointers as well.

A workaround is to wrap the target of map_fn in tf.function:

```
@tf.function  # Fails without this, but shouldn't.
def test_function(x):
    cond = tf.constant(-1)

    if cond == 0:
        result = x
    else:
        result = x

    return result

@tf.function
def partial_call(x):
    tf.map_fn(test_function, x)

X = tf.random.uniform(shape=(10, 224, 224, 3))
partial_call(X)
```

**Review by mdanatg — CHANGES_REQUESTED** [review]
Let's put this test in https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/kernel_tests/map_fn_test.py, since the fix needs to be applied to map_fn.

**Comment by bhack** [comment]
@mdanatg I was posting the same if you see the last commit :smile_cat: 
In the referenced PR yesterday we have solved with the same workaround. 
We could try to solve but do you think that it could be a frequent enough case to fix it?

**Comment by mdanatg** [comment]
It's probably not that frequent, since the functions passed to map_fn are usually simple and don't use control flow. I think it's useful to fix though.

**Review by mdanatg — DISMISSED** [review]
If the test blocks the merge, we can add a `self.skip_test('<explanation or link to issue>')`.

**Comment by bhack** [comment]
@mdanatg Do you have a pointer to the code related to `tf.map_fn` tracing?

**Comment by mdanatg** [comment]
Sure. So map_fn needs to apply autograph to the target function before calling it, here:

https://github.com/tensorflow/tensorflow/blob/8e8c67c3375da3fe8b44e7c11eb1d3fbb2eaa41c/tensorflow/python/ops/map_fn.py#L480

The right way to do that is shown in this code that does the same for Dataset.map:

https://github.com/tensorflow/tensorflow/blob/8e8c67c3375da3fe8b44e7c11eb1d3fbb2eaa41c/tensorflow/python/data/ops/dataset_ops.py#L3223

**Comment by bhack** [comment]
As https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/data/ops/dataset_ops.py#L83 is it not public need I to duplicate
```
autograph_ctx = lazy_loader.LazyLoader(
    "autograph_ctx", globals(),
    "tensorflow.python.autograph.core.ag_ctx")
autograph = lazy_loader.LazyLoader(
    "autograph", globals(),
    "tensorflow.python.autograph.impl.api")
```

**Comment by mdanatg** [comment]
I think it's safest to do that. It might work with a direct import as well, but there are many circular dependencies there.

**Comment by bhack** [comment]
I made a commit but It is going to be a little bit longer then I thought so I cannot go ahead fastly just with colab or blindly changes. 

What is the fast way to have a local setup inside a docker container and to run this changes and the specific test with bazel?

**Comment by mdanatg** [comment]
This is the most practical way I think: https://www.tensorflow.org/install/source#docker_linux_builds.

Once it's working you should be able to run something like `bazel test //tensorflow/python/kernel_tests:map_fn_test`

The initial call to bazel can take a while since it needs to rebuild everything. Afterwards it should be much faster.

**Comment by bhack** [comment]
It was long time ago that I built all from scratch.. I want Google internal bazel cache :smile:

**Comment by bhack** [comment]
I took a very long time to recompile but I don't know cause It seems that the test doesn't fail.

**Comment by bhack** [comment]
Why the test fail only with `@test_util.run_deprecated_v1` decoration and not with `@test_util.run_in_graph_and_eager_modes`?

File: tensorflow/python/kernel_tests/map_fn_test.py
Comment (mdanatg) [inline tensorflow/python/kernel_tests/map_fn_test.py:201]: Use `def_function.function` here (imported as `from tensorflow.python.eager import def_function`)

**Comment by mdanatg** [comment]
Not sure. It's possible that map_fn switches to eager mode even when you have run_in_graph_and_eager_modes. Using tf.function will make sure that's not the case.

File: tensorflow/python/kernel_tests/map_fn_test.py
Comment (bhack) [reply tensorflow/python/kernel_tests/map_fn_test.py]: Ok with this decoration now it is failing correctly

**Comment by bhack** [comment]
I don't think that we could just use `result_value = autograph.tf_convert(fn(elems_value), autograph_ctx.control_status_ctx())`

Cause we get:
``` TypeError: Could not build a TypeSpec for <function convert.<locals>.decorator.<locals>.wrapper at 0x7f651f655e18> with type function```

**Comment by mdanatg** [comment]
> I don't think that we could just use `result_value = autograph.tf_convert(fn(elems_value), autograph_ctx.control_status_ctx())`
> 
> Cause we get:
> ` TypeError: Could not build a TypeSpec for <function convert.<locals>.decorator.<locals>.wrapper at 0x7f651f655e18> with type function`

Correct, tf_convert doesn't call the function, it just translates it.

So we need something like:

```
autographed_fn = autograph.tf_convert(fn, autograph_ctx.control_status_ctx())
result_value = autographed_fn(elems_value)
```

**Comment by bhack** [comment]
Sorry I needed to restarted the container after a system upgrade and I got:
```
INFO: Deleting stale sandbox base /root/.cache/bazel/_bazel_root/68a62076e91007a7908bc42a32e4cff9/sandbox
```
So I will need again hours of recompiling (see https://github.com/tensorflow/build/issues/5 and https://github.com/tensorflow/tensorflow/issues/39560) :( 

Can you use the Google remote or your locale cache to test this a push directly here?

**Comment by mdanatg** [comment]
I can trigger a build which should finish faster.

**Comment by bhack** [comment]
Ok I've done a blind commit. If you can start a CI run.

File: tensorflow/python/ops/map_fn.py
Comment (mdanatg) [inline tensorflow/python/ops/map_fn.py:484]: Remove duplicate line - we must not call the original `fn` again.

**Review by mdanatg — CHANGES_REQUESTED** [review]

File: tensorflow/python/ops/map_fn.py
Comment (bhack) [reply tensorflow/python/ops/map_fn.py]: Done

**Comment by bhack** [comment]
Kokoro we invoke you こころ :pray: :meat_on_bone:

**Comment by bhack** [comment]
I don't know if you can disclose this but is it the Kokoro waiting list only related to the public queue visible with https://github.com/tensorflow/tensorflow/labels/kokoro%3Aforce-run or is there an hidden internal one?

**Review by mdanatg — APPROVED** [review]

**Comment by mdanatg** [comment]
I'm not sure, actually. Usually it picks up immediately.

**Comment by bhack** [comment]
` import/copybara Pending — Waiting for internal safe review approval` is it a manual approval pass right?

**Comment by mdanatg** [comment]
Yep - looks like the tests passed! Consider updating the PR title.

**Comment by bhack** [comment]
> Yep - looks like the tests passed! Consider updating the PR title.

I hope the title could be clear now.

**Pull request merged**
